common/aws/ddb/state.go

```
@@ -238,149 +218,65 @@ func (sdb *StateDatabase) StoreCid(streamCid *models.StreamCid) (bool, error) {
-		var anchoredCid *models.StreamCid = nil
-		if err = sdb.iterateCids(
-			&dynamodb.QueryInput{
-				TableName:              aws.String(sdb.streamTable),
-				IndexName:              aws.String(stateIdAnchorTsIndex),
-				KeyConditionExpression: aws.String("#id = :id and #ancTs >= :ts"),
-				ExpressionAttributeValues: map[string]types.AttributeValue{
-					":id": &types.AttributeValueMemberS{Value: streamId},
-					":ts": &types.AttributeValueMemberN{Value: strconv.FormatInt(streamCid.Timestamp.Unix(), 10)},
-				},
-				ExpressionAttributeNames: map[string]string{
-					"#id":    "id",
-					"#ancTs": "ats",
-				},
-				ScanIndexForward: aws.Bool(true), // ascending
-			},
-			func(streamCid *models.StreamCid) bool {
-				anchoredCid = streamCid
-				return false // stop iteration after the first entry
-			},
-		); err != nil {
-			return nil, err
+		if putItemOut, err := sdb.client.PutItem(ctx, &putItemIn); err != nil {
+			// To get a specific API error
+			var condUpdErr *types.ConditionalCheckFailedException
+			if errors.As(err, &condUpdErr) {
+				// Not an error, just indicate that we couldn't write the entry.
+				return false, nil, nil
+			}
+			log.Printf("updateTip: error writing to db: %v", err)
+			return false, nil, err
+		} else if len(putItemOut.Attributes) > 0 {
+			oldTip := new(models.StreamTip)
+			if err = attributevalue.UnmarshalMapWithOptions(
+				putItemOut.Attributes,
+				oldTip,
+				func(options *attributevalue.DecoderOptions) {
+					options.DecodeTime = attributevalue.DecodeTimeAttributes{
+						S: sdb.tsDecode,
+						N: sdb.tsDecode,
+					}
+				}); err != nil {
+				log.Printf("updateTip: error unmarshaling old tip: %v", err)
+				// We've written the new tip and lost the previous tip here. This means that we won't be able to mark
+				// the old tip REPLACED. As a result, the old tip will just get anchored along with the new tip, which
+				// is ok because we'd rather anchor a few extra tips than miss anchoring some.
```

In what situation would the unmarshall fail?

---

Most likely a bug if/when we change the DynamoDB schema. Shouldn't really happen but I didn't want to just ignore the error.